Added Service Training Module (#9)
WisconsinRobotics/WRoverPlayground: src/wrover_status_light/ServiceIntro.md

# About ROS Services
## What is a ROS Service?

> Services are another form of communication between different software componenents, or more fomally known as nodes, that ROS offers. Common uses of services include sending requests for specific pieces of data, or notify another node to perform specific actions. 

Reviewer: Members should already know what the node structure is
```suggestion
Services are another form of communication between different ROS nodes. Common uses of services include sending requests for specific pieces of data, or notify another node to perform specific actions. 
```

> Services are another form of communication between different software componenents, or more fomally known as nodes, that ROS offers. Common uses of services include sending requests for specific pieces of data, or notify another node to perform specific actions. 

Reviewer: ```suggestion
Services are another form of communication between different software componenents, or more fomally known as nodes, that ROS offers. Common uses of services include sending requests for specific pieces of data, or notify another node to perform specific actions.  A good way to conceptualize a service is as a remote function call.
```

## What is the Makeup of a ROS Service?

ROS services follow a client/server structure, where one node acts as a client that sends a request and another node acts as a server that receives the request and provides a response.

Here is a list of key components involved in services:
* Service definition: services are defined using a srv file. Similar to msg files for ROS topics, the srv file defines the name of a service, the data structure of the request message, and the data structure of the response message. 

> * Client node: the client node in responsible for initiating the service request. It creates the request message, populates it with the required data, and sends it to the server node. the client then waits for the server to process the request and replies with a response.

Reviewer 2: ```suggestion
* Client node: the client node in responsible for initiating the service request. It creates the request message, populates it with the required data, and sends it to the server node. The client then waits for the server to process the request and replies with a response.
```

Author: Addressed in commit [f4eb2b6](https://github.com/WisconsinRobotics/WRoverPlayground/pull/9/commits/f4eb2b6978e8c4699a10c0750eac8bd1e979b5e4).

* Server node: the server node receives and handles the service request calls from client nodes. It retrieves data from the request message, performs the necessary actions, and generates a response message which is sent back to the client. 

> * Request and response messages: these messages are structured data containers that are defined by the service definition. The client and server use these messages to exchange information.

Reviewer 2: Can you break this into 2 bullet points and specify that a request is client-to-server and response is server-to-client?

Reviewer 2: Edits seems to have clarified this at the beginning of the section.

> ## (Optional) Why do we use Services when there's already Topics?

Reviewer: Feels pretty non-optional:
```suggestion
## Why do we use Services when there's already Topics?
```

While both services and topics work to allow communication between nodes, they differ in terms of their data exchange patterns and the types of interactions they support. 

> Unlike publishers and subscribers over topics, which operate asynchronously, service calls provide synchronous communication. The client node waits for the server node to process the request and receive a response before continuing its execution. This synchronous behavior is beneficial in cases where a response is required before proceeding with the next steps of a program.

Reviewer: ```suggestion
Unlike topics, which operate asynchronously, service calls provide synchronous communication. The client node waits for the server node to process the request and receive a response before continuing its execution. This synchronous behavior is beneficial in cases where a response is required before proceeding with the next steps of a program.  It is not acceptable for a server to not exist or to never return a response; these will generate exceptions for the service client to handle.
```

In addition, topics are designed for data streaming and broadcasting continuous streams of information, such as sensor data or robot state updates. They are ideal when nodes require a constant flow of information and do not require immediate responses. On the other hand, services are typically used for accessing functionality that is provided by another node. They are well-suited for situations where nodes need to invoke specific actions or request information from another node, and a response is expected.

## Defining a Service
Services are defined by srv files, which are identical to msg files in terms of formatting, and should be located in the `/srv` folder in a package. In this example, the service takes in a `string` and an `int` as input and returns a `float` as output. 
```
#request fields
string input1
int32 input2
---
#response fields
float32 output
```

> After defining a message, ROS would convert the srv file into three classes: service definition, request messages, and response messages. Suppose we have `ExampleService.srv` in an package called `ExamplePackage`, then the names of these packages would be:

Reviewer 2: Have we explained what packages are at this point in the tutorial? Also, it might be nice to have the information below in a table

Author: Good call about the packages. I think we should cover packages in the first segment of the tutorial, before topics and services. Need to circle back to this later once we begin working on the debugging tools segment.

Author: Added tables in commit [e8b6cb3](https://github.com/WisconsinRobotics/WRoverPlayground/pull/9/commits/e8b6cb3725ccb2e8ef230a247c81e27965bd6d33).

Reviewer 2: Table looks great!

> * Service definition: `ExmaplePackage.srv.ExampleService` for Python and `ExamplePackage::ExampleService` for C++

Reviewer 2: ```suggestion
* Service definition: `ExamplePackage.srv.ExampleService` for Python and `ExamplePackage::ExampleService` for C++
```

> * Request messages: `ExamplePackage.srv.ExampleServiceRequest` for Python and `ExmaplePackage::ExampleService::Request` for C++

Reviewer 2: ```suggestion
* Request messages: `ExamplePackage.srv.ExampleServiceRequest` for Python and `ExamplePackage::ExampleService::Request` for C++
```

* Response messages: `ExamplePackage.srv.ExampleServiceResponse` for Python and `ExamplePackage::ExampleService::Request` for C++

## Writing a Server Node

> ### Python

Reviewer 2: I think adding line numbers here and for the other code examples could make it even better. It might also be good to add whitespace after the description of each code block to visually separate them. Just glancing at it, I wouldn't know whether the description corresponds to the code above or below it.

> A server code can be initailized with

Reviewer 2: ```suggestion
A server code can be initialized with
```

> ```

Reviewer: ```suggestion
``` python
```

Reviewer: Do this for all Python code blocks to get syntax highlighting.

def example_service_server():
    rospy.init_node('example_service_server')
    s = rospy.Service('example_service', ExamplePackage.srv.ExampleService, example_service_execution)
    rospy.spin()

> def example_service_execution(reqest):

Reviewer 2: ```suggestion
def example_service_execution(request):
```

> def example_service_execution(reqest):

Reviewer: Using type annotations on the Python code may make it more explicit what one could do with each variable.
```suggestion
def example_service_execution(reqest: ExampleServiceRequest) -> ExampleServiceResponse:
```

>     return ExamplePackage.srv.ExampleServiceResponse(0.0)

Reviewer: May need to tweak the Python code examples to match the include behavior we need to get autocomplete to work.  Alternatively, you could direct them to the starting files as the example rather than typing it all up a second time.

```
Creates a new service with the name of `example_service` that is defined by the `ExamplePackage.srv.ExampleService` class. When the service is called, the `example_service_execution` method will be called to handle the request. 
```
rospy.spin()
```
Keeps the code running by entering a loop that listens for incoming requests. 

The function `example_service_execution()` is the handler function for this service and is called whenever the node receives a new request message. Handler functions are responsible for carrying out the actual logics of the service and send a response back to the client. Handler functions always take in a single argument of the request message type, and returns an instance of the response message type.

```
return ExamplePackage.srv.ExampleServiceResponse(0.0)
```
In this example, the handler function always returns a hardcoded response of 0.0. If the handler function wants to use the request inputs, they can be accessed with `request.input1` or `request.input2`.

Returning response messages can also be streamlined by not having to manually create a new instance. In this case, simply 
```
return 0.0

> ``` 

Reviewer: This doesn't work: https://wiki.ros.org/rospy/Overview/Services#line-24

Author: Hmm, there seems to be conflicting information, since later on it says this: https://wiki.ros.org/rospy/Overview/Services#line-137
I also tried running my code by just retuning a value instead of a response and it works.

would also work. 

### C++
A server node can be initialized with
```

> bool ExampleServiceServer::callback(ExmaplePackage::ExampleService::Request& request, ExmaplePackage::ExampleService::Response& response) {

Reviewer 2: ```suggestion
bool ExampleServiceServer::callback(ExamplePackage::ExampleService::Request& request, ExamplePackage::ExampleService::Response& response) {
```

    return 0.0;

> }

Reviewer: Why the extra namespace?  This could be a free function, and is probably simpler that way.

> ExmaplePackage::ExampleService exampleService;

Reviewer 2: ```suggestion
ExamplePackage::ExampleService exampleService;
```

> ExmaplePackage::ExampleService exampleService;

Reviewer: ```suggestion
ExamplePackage::ExampleService exampleService;
```

ros::NodeHandle nh;

> ros::ServiceServer service = nh.advertiseService("example_service", &ExampleServiceServer::callblack, &exampleService);

Reviewer 2: ```suggestion
ros::ServiceServer service = nh.advertiseService("example_service", &ExampleServiceServer::callback, &exampleService);
```

```

> To break the code down, the function `ExampleServiceServer::callback()` is the handler function of the server, and is ran every time the server receives a new request. The handler function must take in two arguments, a request and a response.

Reviewer 2: ```suggestion
To break the code down, the function `ExampleServiceServer::callback()` is the handler function of the server, and is run every time the server receives a new request. The handler function must take in two arguments: a request and a response.
```

```

> ExmaplePackage::ExampleService exampleService;

Reviewer 2: ```suggestion
ExamplePackage::ExampleService exampleService;
```

ros::NodeHandle nh;

> ros::ServiceServer service = nh.advertiseService("example_service", &ExampleServiceServer::callblack, &exampleService);

Reviewer 2: ```suggestion
ros::ServiceServer service = nh.advertiseService("example_service", &ExampleServiceServer::callback, &exampleService);
```

> ros::ServiceServer service = nh.advertiseService("example_service", &ExampleServiceServer::callblack, &exampleService);

Reviewer: Still need to `spin`.

```

> The code sets up the service server using a `NodeHandle` instance. NodeHandles provide functions for creating, managing, and interacting using ROS nodes. Here the `NodeHandle` creates a new server with `example_service` as the server name, `ExmapleServiceServer::callback` as the handler function for service calls, and `exmapleService` defines teh service type. 

Reviewer 2: ```suggestion
The code sets up the service server using a `NodeHandle` instance. NodeHandles provide functions for creating, managing, and interacting using ROS nodes. Here the `NodeHandle` creates a new server with `example_service` as the server name, `ExampleServiceServer::callback` as the handler function for service calls, and `exampleService` defines the service type. 
```

## Writing a Client Node
### Python
Services can be called by 
```
rospy.wait_for_service('example_service')
example_service = rospy.ServiceProxy('example_service', ExamplePackage.srv.ExampleService)

> response = example_service(ExamplePackage.srv.ExampleServiceRequest(..., ...))

Reviewer: This segment would also benefit from type annotations.

```
To break the code down,
```
rospy.wait_for_service('exmaple_service')
```
This is a blocking operation that waits until a service with the name `example_service` is created. 
```
example_service = rospy.ServiceProxy('example_service', ExamplePackage.srv.ExampleService)
```

> Creates a new callable proxy object for the service named `example_service` and defined by `ExmaplePackage.srv.ExampleService`.

Reviewer 2: ```suggestion
Creates a new callable proxy object for the service named `example_service` and defined by `ExamplePackage.srv.ExampleService`.
```

Reviewer 2: I think "new callable proxy object" might be difficult for beginners to follow. Maybe just "new object" is enough here.

```

> response = example_service(ExamplePackage.srv.ExmapleServiceRequest(..., ...))

Reviewer 2: ```suggestion
response = example_service(ExamplePackage.srv.ExampleServiceRequest(..., ...))
```

```
Calls the service using an instance of the request message class and saves the response in a local variable. Similar to before, the service can also be called without explicitly making a new instance with
```

> response = exmaple_service(..., ...)

Reviewer 2: ```suggestion
response = example_service(..., ...)
```

```

### C++
Services can be called by
```
ros::NodeHandle nh;
ros::ServiceClient client = nh.serviceClient<Example_Package::Example_Service>("example_service");

Example_Package::Example_Service srv;
srv.request.input1 = ...
srv.request.input2 = ...
if (client.call(srv)) {
    ...
} else {
    ...
}
```
To break the code down
```
ros::NodeHandle nd;
ros::ServiceClient client = nh.serviceClient<Example_Package::Example_Service>("example_service");
```
The code first creates a new NodeHandle instance. Here the NodeHandle instance creates a new ServiceClient using the type `Example_Package::Example_Service` and service name `example_service`.
```
Example_Package::Example_Service srv;
srv.request.input1 = ...
srv.request.input2 = ...
```
The code then creates a new `Example_Service` instance and populates the request variables.
```
if (client.call(srv)) {
    ...
} else {
    ...
}
```
Lastly, the code calls the service using the `call()` function. If the service call is successful, then the response value can be accessed using `srv.response`. If it is unsuccessful, then the code will go into the else block.

## Notes

> ROS's wiki page on services: http://wiki.ros.org/Services

Reviewer: Maybe point to the provided tutorials as well for more example code.

Author: Addressed in commit [cc1f8e6](https://github.com/WisconsinRobotics/WRoverPlayground/pull/9/commits/cc1f8e6a27e5b1774be76c98630bef821db9c083). Let me know if you meant other tutorials.

Services are defined by srv files, which are identical to msg files in terms of formatting, and should be located in the `/srv` folder in a package. In this example, the service takes in a `string` and an `int` as input and returns a `float` as output. You can find an example service file at `srv/DoubleService.srv`.
```
#request fields
string input1
int32 input2
---
#response fields
float32 output
```
After defining a message and building this project, ROS would convert a srv file into three classes: service definition, request messages, and response messages. Suppose we have `ExampleService.srv` in an package called `ExamplePackage`, then the names of these packages would be:
|                   | C++                                     | Python                                    |
|---                 | ---                                     | ---   |
| Service definition |`ExamplePackage::ExampleService`          |`ExamplePackage.srv.ExampleService`        |
| Request message    |`ExamplePackage::ExampleService::Request` |`ExamplePackage.srv.ExampleServiceRequest` |
| Response message   |`ExamplePackage::ExampleService::Response`|`ExamplePackage.srv.ExampleServiceResponse`|

> ## Writing Server and Client nodes

Reviewer 2: A general description of what the demo code does would be nice here. It would also be good to explain the format of the .srv file either in here, or in comments.

Author: Addressed in commit [c71a540](https://github.com/WisconsinRobotics/WRoverPlayground/pull/9/commits/c71a54021d2c6df684d511cbd4e7e1915db39a43).

In addition, topics are designed for data streaming and broadcasting continuous streams of information, such as sensor data or robot state updates. They are ideal when nodes require a constant flow of information and do not require immediate responses. On the other hand, services are typically used for accessing functionality that is provided by another node. They are well-suited for situations where nodes need to invoke specific actions or request information from another node, and a response is expected.

## Defining a Service
Services are defined by srv files, which are similar to msg files in terms of formatting except they contain two parts:
```
#request fields
insert your fields for request messages here as follows:
[data type] [field name]
[data type] [field name]
...
---
#response fields
insert your fields for response messages here as follows:
[data type] [field name]
[data type] [field name]
...
```

> Fields are listed in the srv file each on its individual line, with a `---` separating the request fields from the response fields. Both requests and responses can have a single field, multiple fields, or no fields at all. All srv files for a module should be located in the `/srv` folder for proper complication. 

Reviewer 2: Could you also explain here how constants could be defined? I see that in ServiceLight.srv, DONE and NOT_DONE are defined. I'm assuming these are used as constants which can be assigned to the request field?

Author: Added in [a470552](https://github.com/WisconsinRobotics/WRoverPlayground/pull/9/commits/a47055254bc5a7390410e548bba23e97ab80c470).

#response fields
float32 output
```
After defining a message and building this project, ROS would convert a srv file into three classes: service definition, request messages, and response messages. Suppose we have `ExampleService.srv` in an package called `ExamplePackage`, then the names of these packages would be:
|                   | C++                                     | Python                                    |
|---                 | ---                                     | ---   |
| Service definition |`ExamplePackage::ExampleService`          |`ExamplePackage.srv.ExampleService`        |
| Request message    |`ExamplePackage::ExampleService::Request` |`ExamplePackage.srv.ExampleServiceRequest` |
| Response message   |`ExamplePackage::ExampleService::Response`|`ExamplePackage.srv.ExampleServiceResponse`|

## Writing Server and Client nodes
On how to write server and client nodes, see the demo files in the `src` folder for example code and explanations. Come back to this file and continue when you are done going through them. 

## Running Server and Client nodes
### Python

### C++

## Challenge

> **TODO**

Reviewer: TODO: Fix this TODO

Author: Addressed in commit [3884016](https://github.com/WisconsinRobotics/WRoverPlayground/pull/9/commits/38840160b9024e455aeda539c1a47f9134ce7f07).

| Response message   |`ExamplePackage::ExampleService::Response`|`ExamplePackage.srv.ExampleServiceResponse`|

## Writing Server and Client nodes
Inside this module are two demo code files in the `src` folder, one in Python and one in C++. The server code creates a service server node that takes in an interger and doubles it. The client code creates a service client node that uses the server to double a number on repeat. Note that since the service uses `int64` (a integer with 64 bits), running the client code will eventually reach an integer overflow. 

Read through the demo files to understand how to write server and client nodes. Come back to this file and continue when you are ready.  

### C++ Server
A basic server node can be initialized in C++ with:
``` C++
ros::init(argc, argv, "double_server");
ros::NodeHandle nh;
ros::ServiceServer service = nh.advertiseService("double_service", doubleServiceCallback);
ros::spin();

> bool doubleServiceCallback(wrover_status_light::DoubleServiceRequest& req, wrover_status_light::DoubleServiceResponse& res) {

Reviewer 2: In C++ this would have to be declared before it used in the call to nh.advertiseService(...). For this reason, consider moving this function definition to the top of the example.

Author: Addressed in [c0595c3](https://github.com/WisconsinRobotics/WRoverPlayground/pull/9/commits/c0595c34d291b4b911ab00f2ad07d58d9c26804b).

### Python Client
A basic client node can be initialized in Python with: 
``` Python
rospy.init_node("service_client")
double_service = rospy.ServiceProxy("double_service", DoubleService)
rospy.wait_for_service("double_service")
response:DoubleServiceResponse = double_service(num)
```
To break the code down: 
``` Python
rospy.init_node("service_client")
```
This creates a new node called `service_client`.
``` Python
double_service = rospy.ServiceProxy("double_service", DoubleService)
```
This creates a handle to call the service. Here we specify the name of the service, `double_service`, and the type of the service, `DoubleService`. 
``` Python
rospy.wait_for_service("double_service")
```
This waits until the service is created and ready. 
``` Python
response:DoubleServiceResponse = double_service(...)
```
This calls the service with a request message as input, and saves the returned response message. 

## Running Server and Client nodes
### C++
To run the C++ code:
1. Navigate to this repository using `cd {your file path}/WRoverPlayground`.
2. Run `catkin_make` to build the project if you haven't yet.

> 3. Run `source devel/setup.bash` to source the project if you haven't yet. Also keep in mind that you have to resource every time you open a new terminal. 

Reviewer 2: ```suggestion
3. Run `source devel/setup.bash` to source the project if you haven't yet. Also keep in mind that you have to rerun the command every time you open a new terminal. 
```

Reviewer 2: "resource" is also a different word, so might be confusing :)

4. Run `roscore` to start ROS master.
5. To run the server code, in a new terminal, run `rosrun wrover_status_light DemoServer`.
6. To see a list of all services, in a new terminal, run `rosservice list`. You should see `/double_service` listed. Now would also be a good opportunity to run `rosservice help` and familiarize yourself with all the `rosservice` commands.
7. You can run the service manually from the command line using `rosservice call`. In this case, try running `rosservice call /double_service "input: [insert your number here]"`. You should see that the inputted number is doubled. 
8. To run the client code, run `rosrun wrover_status_light DemoClient`. You should see powers of 2 being printed. Eventually the program will print `Integer overflow encountered`. This is because the response/request messages use a 64-bit integer, and as the number is doubled eventually we will reach integer overflow.
9. Use `Ctrl-C` to exit out of any programs

### Python
To run the Python code:
1. Navigate to this repository using `cd {your file path}/WRoverPlayground`.
2. Run `catkin_make` to build the project if you haven't yet.

> 3. Run `source devel/setup.bash` to source the project if you haven't yet. Also keep in mind that you have to resource every time you open a new terminal. 

Reviewer 2: ```suggestion
3. Run `source devel/setup.bash` to source the project if you haven't yet. Also keep in mind that you have to rerun the command every time you open a new terminal. 
```

Reviewer 2: "resource" is also a different word, so might be confusing :)
